### ISSUE TITLE
The message doesn't get processed if a message "frame" ends with ":["

### ISSUE THREAD

**Issue by KushtrimPacaj** [SignalR/java-client#104]
I was having an issue while using this library that sometimes I would invoke a hub method in my android client, but I wouldn't receive a callback at all ( neither Done or onError ). Other methods were working fine, but that specific method with those specific parameters wouldn't.

I started investigating, and finally tracked the culprit to this part of the library:  https://github.com/SignalR/java-client/blob/98ae5a47426da4d416a896c547080797bbafefae/signalr-client-sdk/src/main/java/microsoft/aspnet/signalr/client/transport/WebsocketTransport.java#L122

The problem was that since I was transmitting JSON, sometimes a frame of my message would end with ":["   ( which is the start of an array in JSON). The library then would detect the frame of my message as a prefix, and try to prepend it as such to the other frames. This would most of the time produce an invalid json, and the whole message would never get processed.  

I know that there are workarounds that I can do, somehow encode my message differently, but I think that the way that we check for a prefix is inherently wrong in the library, and should be modified.

**Comment by KushtrimPacaj** [comment]
@jasminsehic I noticed that it was you who wrote the class. Is there a standard format for the message prefix, so that we can match it with regex or something like that ?

In my tests, the message is always in this format : `{ "C":"blablabla","I": 42 ,"M":[` but I can't be sure that this is the only valid prefix format. Could you elaborate ?

**Comment by MinChanSike** [comment]
@KushtrimPacaj  I also got that issue, callback not came out after invoked my method.
In my tests, callback not response if response data size is more than _2KB_.
Finally I got  `invalid json received:` error message.

**Comment by WhyTree-KushtrimPacaj** [comment]
@MinChanSike 
About you issue, have you seen #75 ? If you receive more than 2 frames ( prefix + payload ), you have to append them manually.
That is not related to this issue when a wrong prefix is selected if a data frame ends with ":["

**Comment by KushtrimPacaj** [comment]
@MinChanSike 
About you issue, have you seen #75 ? If you receive more than 2 frames ( prefix + payload ), you have to append them manually.
That is not related to this issue when a wrong prefix is selected if a data frame ends with ":["

**Comment by MinChanSike** [comment]
Thanks _Kushtrim_. 
I try and solve it by the following. 
What can impact from my fixed?
As my testing, I have detect message start with `{"R":` and end with `"I":"` .
How about in your issue? It can any help? 

```C#
private String mContent;
public void onFragment(Framedata frame) {
                try {
                    String decodedString = Charsetfunctions.stringUtf8(frame.getPayloadData());                  
                 if (decodedString.contains("{\"R\":") || null == mPrefix) {
                        mPrefix = decodedString;
                        return;
                    }
                    mContent = (mContent == null) ? decodedString : mContent + decodedString;
                    if (decodedString.contains(",\"I\":\"") && mContent != null) {
                        String simpleConcatenate = mPrefix + mContent;
                        if (isJSONValid(simpleConcatenate)) {
                            onMessage(simpleConcatenate);
                        } else {
                            String extendedConcatenate = simpleConcatenate + "]}";
                            if (isJSONValid(extendedConcatenate)) {
                                onMessage(extendedConcatenate);
                            } else {
                                log("Invalid json received -> " + decodedString, LogLevel.Critical);
                            }
                        }
                    }

                } catch (InvalidDataException e) {
                    e.printStackTrace();
                }
            }
        };
```

**Comment by KushtrimPacaj** [comment]
Check my previous comment, that is what normally gets processed as prefix. Your code won't process that because it doesn't contain a {"R":

**Comment by mrpoetra22** [comment]
Hi you can try this.


/*
Copyright (c) Microsoft Open Technologies, Inc.
All Rights Reserved
See License.txt in the project root for license information.
*/

package microsoft.aspnet.signalr.client.transport;

import java.io.UnsupportedEncodingException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URLEncoder;

import com.google.gson.Gson;

import org.java_websocket.client.WebSocketClient;
import org.java_websocket.exceptions.InvalidDataException;
import org.java_websocket.framing.Framedata;
import org.java_websocket.handshake.ServerHandshake;
import org.java_websocket.util.Charsetfunctions;

import microsoft.aspnet.signalr.client.ConnectionBase;
import microsoft.aspnet.signalr.client.LogLevel;
import microsoft.aspnet.signalr.client.Logger;
import microsoft.aspnet.signalr.client.SignalRFuture;
import microsoft.aspnet.signalr.client.UpdateableCancellableFuture;
import microsoft.aspnet.signalr.client.http.HttpConnection;

/**
 * Implements the WebsocketTransport for the Java SignalR library
 * Created by stas on 07/07/14.
 */
public class WebsocketTransport extends HttpClientTransport {

    private StringBuilder mPrefix;
    private static final Gson gson = new Gson();
    WebSocketClient mWebSocketClient;
    private UpdateableCancellableFuture<Void> mConnectionFuture;

    public WebsocketTransport(Logger logger) {
        super(logger);
    }

    public WebsocketTransport(Logger logger, HttpConnection httpConnection) {
        super(logger, httpConnection);
    }

    @Override
    public String getName() {
        return "webSockets";
    }

    @Override
    public boolean supportKeepAlive() {
        return true;
    }

    @Override
    public SignalRFuture<Void> start(ConnectionBase connection, ConnectionType connectionType, final DataResultCallback callback) {
        final String connectionString = connectionType == ConnectionType.InitialConnection ? "connect" : "reconnect";

        final String transport = getName();
        final String connectionToken = connection.getConnectionToken();
        final String messageId = connection.getMessageId() != null ? connection.getMessageId() : "";
        final String groupsToken = connection.getGroupsToken() != null ? connection.getGroupsToken() : "";
        final String connectionData = connection.getConnectionData() != null ? connection.getConnectionData() : "";


        String url = null;
        try {
            url = connection.getUrl() + "signalr/" + connectionString + '?'
                    + "connectionData=" + URLEncoder.encode(URLEncoder.encode(connectionData, "UTF-8"), "UTF-8")
                    + "&connectionToken=" + URLEncoder.encode(URLEncoder.encode(connectionToken, "UTF-8"), "UTF-8")
                    + "&groupsToken=" + URLEncoder.encode(groupsToken, "UTF-8")
                    + "&messageId=" + URLEncoder.encode(messageId, "UTF-8")
                    + "&transport=" + URLEncoder.encode(transport, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }

        mConnectionFuture = new UpdateableCancellableFuture<Void>(null);

        URI uri;
        try {
            uri = new URI(url);
        } catch (URISyntaxException e) {
            e.printStackTrace();
            mConnectionFuture.triggerError(e);
            return mConnectionFuture;
        }

        mWebSocketClient = new WebSocketClient(uri) {
            @Override
            public void onOpen(ServerHandshake serverHandshake) {
                mConnectionFuture.setResult(null);
            }

            @Override
            public void onMessage(String s) {
                callback.onData(s);
                mPrefix = null;
            }

            @Override
            public void onClose(int i, String s, boolean b) {
                mWebSocketClient.close();
            }

            @Override
            public void onError(Exception e) {
                mWebSocketClient.close();
            }

            @Override
            public void onFragment(Framedata frame) {
                try {
                    String decodedString = Charsetfunctions.stringUtf8(frame.getPayloadData());

                    if(decodedString.equals("]}")){
                        return;
                    }

                    if(decodedString.endsWith(":[") || null == mPrefix){
                        mPrefix = new StringBuilder();
                        mPrefix.append(decodedString);
                        return;
                    }

                    mPrefix.append(decodedString);
                    String simpleConcatenate = mPrefix.toString();


                    if (!isJSONValid(simpleConcatenate)) {
                        if (isJSONValid(simpleConcatenate +"]}")){
                            onMessage(simpleConcatenate +"]}");
                            return;
                        }
                        return;
                    }

                    onMessage(simpleConcatenate);
                } catch (InvalidDataException e) {
                    e.printStackTrace();
                }
            }
        };
        mWebSocketClient.connect();

        connection.closed(new Runnable() {
            @Override
            public void run() {
                mWebSocketClient.close();
            }
        });

        return mConnectionFuture;
    }

    @Override
    public SignalRFuture<Void> send(ConnectionBase connection, String data, DataResultCallback callback) {
        mWebSocketClient.send(data);
        return new UpdateableCancellableFuture<Void>(null);
    }

    private boolean isJSONValid(String test){
        try {
            gson.fromJson(test, Object.class);
            return true;
        } catch(com.google.gson.JsonSyntaxException ex) {
            return false;
        }
    }
}

**Comment by MinChanSike** [comment]
Hi @KushtrimPacaj 
In my tests, I see this message template.
If the server reply **_String_** data, I got
`{"R":"1234567891011121314151617181920","I":"0"}`

If the server reply **_String Array_** data, I got
`{"R":["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","I":"0"}`

If the server reply **_Custom-Object_** data, I got
`{"R":{"ObjId":1,"ObjName":"TestObject1","Field1":"tesging1","AccessRight":{"Control":false,"View":true}},"I":"0"}`

If the server reply **_Custom-Object List_** data, I got
```
{"R":[{"ObjId":1,"ObjName":"TestObject1","Field1":"tesging1","Right":{"Control":false,"View":true}},
{"ObjId":2,"ObjName":"TestObject2","Field1":"tesging2","Right":{"Control":false,"View":true}},
{"ObjId":3,"ObjName":"TestObject3","Field1":"tesging3","Right":{"Control":false,"View":true}}],"I":"0"}
```

My SignalR server is using _Microsoft.AspNet.SignalR_ version _2.1.2_.
This is fixed for me.

```Java
	private int msgFrame = 0;
    private String mContent;
	@Override
	public void onFragment(Framedata frame) {
		try {
			String decodedString = Charsetfunctions.stringUtf8(frame.getPayloadData());
			msgFrame += 1;
			if (decodedString.contains("{\"R\":") || null == mPrefix) {
				mPrefix = decodedString;
				log("Frame:"+ msgFrame +", PrefixData -> " + mPrefix, LogLevel.Information);
				return;
			}
			mContent = (mContent == null) ? decodedString : mContent + decodedString;
			log("Frame:"+ msgFrame +", Content -> " + mContent, LogLevel.Information);
			if (decodedString.contains(",\"I\":\"")) {
				String simpleConcatenate = mPrefix + mContent;
				log("Last Frame:"+ msgFrame +", FullContent -> " + simpleConcatenate , LogLevel.Information);
				if (isJSONValid(simpleConcatenate)) {
					onMessage(simpleConcatenate);
				} else {
					String extendedConcatenate = simpleConcatenate + "]}";
					if (isJSONValid(extendedConcatenate)) {
						onMessage(extendedConcatenate);
					} else {
						String errorMsg = "Invalid json received -> " + decodedString;
						log(errorMsg, LogLevel.Critical);
						onError(new Exception(errorMsg));
					}
				}
				mContent = "";
			}
		} catch (InvalidDataException e) {
			e.printStackTrace();
			mContent = "";
		}
	}
};
```
Thanks for discuss. How about your issue? Can I see your solution?

**Comment by KushtrimPacaj** [comment]
Like I said in the previous comment, as far as I can tell the valid prefix is of this format `{"C":"s-0,431437","M":[`, since this is the only payload I receive that has SignalR metadata and ends with `:[`.

For the moment, as a workaround, I've added a check so that it's a prefix if it ends with `:[` and is the first payload.
Sometimes this prefix doesn't come with the payload at all, in that case the response part comes first. So I added another check to not have it as prefix if it contains "R".

So for the moment my code looks like this:

```
if((decodedString.endsWith(":[") && isFirstFrame && !decodedString.contains("\"R\"")) || null == mPrefix){
                        mPrefix = decodedString;
                        return;
                    }
```

But all these are all just based in tests and logs, and without analyzing the whole SignalR .Net part we can't be sure that I've covered all the use cases. So maybe we should do that...
